test/e2e/framework/util.go

```diff
@@ -2066,15 +2066,32 @@ func AssertCleanup(ns string, selectors ...string) {
 	if ns != "" {
 		nsArg = fmt.Sprintf("--namespace=%s", ns)
 	}
-	for _, selector := range selectors {
-		resources := RunKubectlOrDie("get", "rc,svc", "-l", selector, "--no-headers", nsArg)
-		if resources != "" {
-			Failf("Resources left running after stop:\n%s", resources)
-		}
-		pods := RunKubectlOrDie("get", "pods", "-l", selector, nsArg, "-o", "go-template={{ range .items }}{{ if not .metadata.deletionTimestamp }}{{ .metadata.name }}{{ \"\\n\" }}{{ end }}{{ end }}")
-		if pods != "" {
-			Failf("Pods left unterminated after stop:\n%s", pods)
+
+	backoff := wait.Backoff{
+		Duration: 5 * time.Second,
+		Factor:   2,
+		Steps:    3,
+	}
+	var e error
+	verifyCleanupFunc := func() (bool, error) {
+		e = nil
+		for _, selector := range selectors {
+			resources := RunKubectlOrDie("get", "rc,svc", "-l", selector, "--no-headers", nsArg)
+			if resources != "" {
+				e = fmt.Errorf("Resources left running after stop:\n%s", resources)
```

Comment:
In both cases here and few lines below you can just return the error with the false. This way the error will bubble up to `wait.ExponentialBackoff` caller. It'll be either your error or the timeout error returned by the wait, but in both cases you'll have a clear information what happened, better than the current.

Reply:
> you can just return the error with the false.

If you take a closer look at `wait.ExponentialBackoff`, you will see it won't retry at all when the condition function returns an error.

Reply:
@soltysh  In my previous change, the error was returned with false . Then Mengqi pointed out that `wait.ExponentialBackoff` will exit immediately when it sees an error, which causes the retry end early.